Fix stream syslog socket delimiters, reconnect on stream socket failures, use RFC 5424 format

jikamens opened the pull request:
Fixes #241.

jikamens commented:
Sorry about all the crap commits. I've squashed them into the actual functional commits and repushed.

jikquantopian commented:
Ping?

vmalloc commented on logbook/handlers.py:
```diff
@@ -1391,14 +1391,26 @@ def __init__(self, application_name=None, address=None,
                 address = '/dev/log'
 
-        self.address = address
+        self.remote_address = self.address = address
```
Why is `remote_address` needed? Why not just use `address`?

jikamens replied:
This is explained in the commit message for b9daef7311f4816467db8af2b7573bad1bb55c33.

vmalloc commented on logbook/handlers.py:
```diff
@@ -1391,14 +1391,26 @@ def __init__(self, application_name=None, address=None,
+
+        try:
+            self.connection_exception = BrokenPipeError
+        except NameError:
```
This is a dangerous way of checking if BrokenPipeError is supported.

jikamens replied:
Why? I don't see what's dangerous about it.

Can you suggest an alternative that is as portable?

vmalloc replied:
Catching `NameError` is considered dangerous because this would also raise it:

```
sellf.connection_exception = BrokenPipeError
```
And also this (which is a common typo for some folks):
```
shelf.connection_exception = ...
```
Also, at some point in the future `self` could implement `__setattr__` or `__setattribute__` which themselves could introduce the exception, and this would quietly swallow it.

A better way is to use
```
getattr(__builtin__, 'BrokenPipeError', <fallback>)
```

vmalloc commented:
Sorry - was a bit busy lately so didn't get around to it. I commented in two places for clarifications.

Thanks!

vmalloc commented:
It also seems the Travis tests are failing... have any idea why?

jikamens commented:
> It also seems the Travis tests are failing... have any idea why?

I didn't want to spend time working on the unit tests until I know y'all were comfortable with these changes conceptually. Let me know if you are, and I'll dig into why the tests are failing and whether fixes are needed to the tests or to my proposed changes.

vmalloc commented:
@jikamens I think I'm fine with the changes as long as they don't break existing (reasonable) usage, and as long as they make sense logically...

Having said that, once we merge and release it, we could be seeing issues from people already using Logbook having issues with the change. Will you be available to aid in fixes? I have little time to actively commit to this project nowadays...

jikamens commented:
OK, all issues you mentioned are addressed, I added a couple bonus fixes, I rebased against current getlogbook/master, and Travis is passing. And yes, I'll be around to help if there are bug reports or complaints. Please review, and if everything looks good I'll squash and rebase one more time before merge.

jikamens commented:
Ok, rebased and added changelog entries, though I didn't bump the version number since I don't know what version number you want to use when you release these changes. I put "Not yet released" in the CHANGES file instead of specifying a version number.

vmalloc commented:
Looks good! Thanks!
